pkg/tfshim/sdk-v2/provider2.go

```diff
@@ -833,8 +831,53 @@ func normalizeBlockCollections(val cty.Value, res *schema.Resource) cty.Value {
 			} else {
 				contract.Failf("normalizeBlockCollections: Unexpected field type %v", fieldType)
 			}
+		} else {
+			subBlockSchema := res.SchemaMap()[fieldName]
+			if subBlockSchema == nil {
+				contract.Failf("normalizeBlockCollections: Unexpected nil subBlockSchema for %s", fieldName)
+			}
+			subBlockRes, ok := subBlockSchema.Elem.(*schema.Resource)
+			if !ok {
+				contract.Failf("normalizeBlockCollections: Unexpected schema type %s", fieldName)
+			}
+
+			subBlockVal := val.GetAttr(fieldName)
+			if subBlockVal.Type().IsListType() {
+				newSlice := normalizeIterable(subBlockVal, subBlockRes)
+				if len(newSlice) != 0 {
+					valMap[fieldName] = cty.ListVal(newSlice)
+				} else {
+					valMap[fieldName] = cty.ListValEmpty(subBlockVal.Type().ElementType())
+				}
+				continue
+			} else if subBlockVal.Type().IsSetType() {
+				newSlice := normalizeIterable(subBlockVal, subBlockRes)
+				if len(newSlice) != 0 {
+					valMap[fieldName] = cty.SetVal(newSlice)
+				} else {
+					valMap[fieldName] = cty.SetValEmpty(subBlockVal.Type().ElementType())
+				}
+			} else {
+				contract.Failf("normalizeBlockCollections: Unexpected field type %v", subBlockVal.Type().GoString())
+			}
+
+			valMap[fieldName] = normalizeBlockCollections(val.GetAttr(fieldName), subBlockRes)
+
 		}
 	}
-	// TODO nested blocks?
 	return cty.ObjectVal(valMap)
 }
+
+func normalizeIterable(blockVal cty.Value, blockRes *schema.Resource) []cty.Value {
+	if blockVal.IsNull() ||
```

Comment:
there is the interesting question of unknowns. You may be assuming here that MakeTerraformInputs has taken care of unknowns by this point, but if so it's good to assert. Otherwise will just panic in  blockVal.AsValueSlice() unhepfully.

Reply:
Good point, thanks!